### ISSUE TITLE
Node Coloring

### ISSUE THREAD

**Issue by eileenchoe** [dondi/GRNsight#566]
Nodes should be colored to visualize the gene expression data extracted from the GRNmap input or output workbook with #241.

Some initial design considerations:
- Color scheme should not conflict with the cyan / magenta edges for weighted networks
- Coloring should not obscure the gene name inside the node
- Corresponding UI menu should be added to toggle/adjust node coloring options
  - User should be able to chose which strain (e.g. wt, dgln3, etc.) to use for expression viz
- If experimental AND simulated expression data are present in the output worksheet, visualization should facilitate visual comparison. Experimental data are in `*_expression` sheets, while simulated expression data is stored in `*_optimized_expression` sheets.

**Comment by eileenchoe** [comment]
<img src="https://user-images.githubusercontent.com/16827883/35782997-96b7e014-09b5-11e8-8398-1bcb6991a0b2.JPG" width="200">
Here's a sketch of ideas for node coloring methods before I start prototyping with d3.

(1) and (4) are options for visualizing experimental data only
(2) (3) and (5) are options for visualizing both experimental and simulated data.

I'm leaning toward (4) and (5) as they offer good readability of the gene name.

**Comment by eileenchoe** [comment]
Bookmarking this link for [handy built in methods for data structures in D3-collections](https://github.com/d3/d3-collection).

**Comment by eileenchoe** [comment]
Looked into the [d3-chromatic library](https://github.com/d3/d3-scale-chromatic#diverging) which offers diverging color schemes (like red -> blue with range [0,1]) which can be continuously interpolated or discretized between 3-11 color "bins." 

Normalization:
- Expression values will have to be normalized such that 0.5 on the color scale corresponds to 0 for expression value.
- The `[-/+]abs(max(expression_values))` should "map to" the 0 and 1 endpoints of the color scale.
- Thus, if expression values were between [-1.34, .45] it's corresponding color mapping should normalize and scale the "positive range" [0 - 1.34) => [0.5 - 1]

**Comment by kdahlquist** [comment]
Need to run the colors by some color blind users.  @kdahlquist thinks that we will choose either pink/green, purple/teal, or red/blue.

Option numbers 1 and 3 for the node coloring are the ones we need.

In terms of normalization, the danger is that for graphs where all genes have low expression values, the user will see bright colors and assume that the genes are actually changing more than they are.  Some options for the user:
- normalize based on the absolute value of the largest magnitude fold change
- allow user to set normalization factor (like we do for edges)
- have a "preset" normalization factor, like 10 for the largest magnitude 

mapping of expression value to color scheme is linear

GenMAPP used tick-marks to separate the stripes.  For now, we will not have any, but could add in later

Probably will change colors of edges to match the colors chosen for expression data.

Later we could have options for different color choices, but for now will stick with one.

**Comment by kdahlquist** [comment]
Number of bins?  Answer later.

**Comment by eileenchoe** [comment]
Link to Issue on GRNmap project regarding diverging color schemes that are friendly for color blind users: https://github.com/kdahlquist/GRNmap/issues/396

**Comment by eileenchoe** [comment]
- Got an initial pass of node coloring working! Biggest hurdle was figuring out how to associate the correct expression data array to each node. I found a d3 method `.each()` that I used to iterate over the nodes.
- Here’s a screenshot of node coloring for “dcin5_log2_expression” strain data, with the normalization factor set to 3. I am using the “RdBu” diverging color scheme from d3-scale-chromatic, using a continuous scale (will change later to discrete using 10 “bins” as we discussed). Thus, the expression values [-3, 3] are mapping to [Blue, Red]. This is visually confirmed by the fact that, for example, ASH1 is all different shades of blue, and its data in “dcin5_log2_expression” is [-1.3229	-0.4236	-1.8945	-1.6929	-2.2704	-1.0939	-1.3098	-1.017	-1.7106	-1.7777-1.9029	-1.3284]

<img width="1280" alt="screen shot 2018-02-23 at 5 25 40 pm" src="https://user-images.githubusercontent.com/16827883/36624096-02245436-18c0-11e8-8532-b1dc0a904482.png">

- There’s a lot of values hardcoded for now. Over the next week I will work on writing functions for dynamically sizing the node coloring slices, and simple UI elements to allow users to customize node coloring parameters (expression strain and normalization factor seem to be the highest priority for this).

- Lastly, we allow for missing data in the expression data, what color (if any) should null data map to?

**Comment by kdahlquist** [comment]
Null data should be gray or white, something that is unobtrusive and not on the scale.

I'm only looking at this on my phone, but this looks really good. 

Between you and Jen, we really need to have a talk about reorganizing the menu and ui.

**Comment by eileenchoe** [comment]
<img width="1280" alt="screen shot 2018-03-12 at 5 34 42 pm" src="https://user-images.githubusercontent.com/16827883/37316364-ce7476fc-261b-11e8-8133-322b3c5d8f4f.png">
Got top and bottom dataset selection working, as well as changed the edge color scheme to the second-to-strongest colors on each side of the spectrum on the legend!

**Comment by eileenchoe** [comment]
TODO:
- [ ] Write logic for the bottom dataset to remain the "Same as Top Dataset" unless explicitly changed
- [ ] Change the node coloring legend to be continuous, and only include 3 values: left, 0, and right

**Comment by eileenchoe** [comment]
<img width="1280" alt="screen shot 2018-03-31 at 5 07 51 pm" src="https://user-images.githubusercontent.com/16827883/38168613-338c7082-3506-11e8-8aa5-11ad363c74c6.png">
Completed some of the features (checked off on TODO list above), @kdahlquist what do you think of the checkbox for compute averages and the button toggle for enable/disable node coloring?

**Comment by kdahlquist** [comment]
- [ ] Drop down menus need to be resized so that they don't overlap the bounding box.  Long sheet names are going to have to get truncated with a ... extension to indicate it was truncated.  @eileenchoe will look into whether it is possible to show the full sheet name in the drop down once the user clicks on it.
- [ ] Remove the word "menu" after "Node Coloring" because we don't use the word menu anywhere else in the control panel.
- [ ] Move the enable/disable button to immediately below the "Node Coloring" header
- [ ] We need check boxes for averaging values for both the top and bottom datasets because the user might decide to compare the average to the raw data for the same sheet.
- [ ] Instead of "Average Values" say "Average Replicate Expression Values"
- [ ] Legend should be in viewport, either top left or bottom right; need option to show/hide legend

**Comment by eileenchoe** [comment]
<img width="1280" alt="screen shot 2018-04-04 at 7 39 24 pm" src="https://user-images.githubusercontent.com/16827883/38344729-fb09409a-383f-11e8-8037-4e68d8330886.png">
We've got grid layout + node coloring! Left menu bar is getting pretty long..

**Comment by eileenchoe** [comment]
<img width="226" alt="screen shot 2018-04-06 at 2 31 16 pm" src="https://user-images.githubusercontent.com/16827883/38445321-e630cbe8-39a7-11e8-91dc-03016ff31e89.png">
<img width="1280" alt="screen shot 2018-04-06 at 2 31 12 pm" src="https://user-images.githubusercontent.com/16827883/38445324-e7eea1d0-39a7-11e8-8ed0-ca8745f87bf6.png">



Incorporated most of the feedback above. Arbitrarily chose 24 characters as the max # of characters before the sheet name is truncated with "..." to fit in the left panel. Investigated a little bit into dynamic resize to expand truncated sheet names, but in the interest of time, I'll write another issue for this to be completed later. 

"Average Replicate Expression Values" does not fit in the panel. "Avg Replicate Expression Vals" fits.

Note: Node coloring will only be applied for sheets whose names end with "expression," and the node coloring menu automatically appears when a Workbook with at least 1 expression sheet is uploaded.

Will be doing code cleanup this weekend before submitting the PR.

**Comment by kdahlquist** [comment]
How about "Average Replicate Values" and leave out "expression".

Let's target talking about the side panel on Monday.  @eileenchoe and @yshin4 need to work on the menu changes we discussed in a previous meeting.

**Comment by eileenchoe** [comment]
Realized that in order for node coloring settings not to reset upon reload, I need to factor node coloring out of `graph.js` into a node coloring controller object that is instantiated and passed into drawGraph function in `upload.js`. This allows the node coloring to save state upon reload. Related to #576.

**Comment by eileenchoe** [comment]
Completed refactor, and PR #582 has been submitted for this issue!

**Comment by eileenchoe** [comment]
Also completed the addition of node coloring features to featureList.json for client side testing document, and it is included in PR #582.

**Comment by eileenchoe** [comment]
Live on beta v.2.3.6
I did not get a chance to complete time point legend & color palette selector features, which were lower priority. I will make new, separate issues for these two features shortly.

**Comment by eileenchoe** [comment]
#583 and #584 document the two features referenced above that were not completed in this issue.

**Comment by kdahlquist** [comment]
What are the max and minimum allowable values that a user can type in the Log Fold Change Max Value field?  I typed a really big number (like 1000000) and it seemed to slow down the browser.  

If there is no max, then we should choose one.  I think 100 (and -100) is sufficiently large for this.

**Comment by kdahlquist** [comment]
Cosmetic comment: might as well make the "Enable Node Coloring" button the full width of the area like the other buttons, even though the text label is not that long.

**Comment by kdahlquist** [comment]
Question for @eileenchoe: how are you dealing with missing expression data? That is,
- Can GRNsight deal with it without throwing an error?
- What color is assigned to it?

**Comment by eileenchoe** [comment]
@kdahlquist missing expression data does not throw an error, but is converted to 0 on the color scale as discussed in a comment above.

**Comment by kdahlquist** [comment]
OK, thanks for clarifying.

**Comment by eileenchoe** [comment]
Discovered that the text blurring was due to not clearing the node text prior to redrawing it, so this was resolved without a separate redraw option. 

Implemented (-100 - 100) as range for log fold change value. I realize that in beta, the node coloring legend does not render for negative numbers so I also fixed this.

I can implement Set Button for normalization factor. Just an FYI -- @yshin4 and I are not implementing set buttons for any of the input boxes that appear in the menu dropdowns. They are triggered with clicking enter.

Changes will be folded into menu updates #462 PR
